com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -0,0 +1,343 @@
+         *            debug source attribute, or <code>null</code> if none
+         */
+        private IResource getSourceFile(IJavaProject project, String qualifiedName, String sourceAttribute) {
+            String name = null;
+            IJavaElement element = null;
+            try {
+                if (sourceAttribute == null) {
+                    element = findElement(qualifiedName, project);
+                } else {
+                    int i = qualifiedName.lastIndexOf('/');
+                    if (i > 0) {
+                        name = qualifiedName.substring(0, i + 1);
+                        name = name + sourceAttribute;
+                    } else {
+                        name = sourceAttribute;
+                    }
+                    element = project.findElement(new Path(name));
+                }
+                if (element instanceof ICompilationUnit) {
+                    ICompilationUnit cu = (ICompilationUnit) element;
+                    return cu.getCorrespondingResource();
+                }
+            } catch (CoreException e) {
+                logger.log(Level.INFO, "Failed to get source file with exception" + e.getMessage(), e);
+            }
+            return null;
+        }
+    }
+
+    class HotCodeReplaceEvent extends DebugEvent {
+
+        public String message;
+
+        /**
+         * Constructor.
+         */
+        public HotCodeReplaceEvent(String message) {
+            super("hotCodeReplace");
+            this.message = message;
+        }
+    }
+
+    @Override
+    public void initialize(IDebugAdapterContext context, Map<String, Object> options) {
+        // Listen to the built file events.
+        ResourcesPlugin.getWorkspace().addResourceChangeListener(this, IResourceChangeEvent.POST_BUILD);
+
+        // TODO: Change IProvider interface for shutdown event
+    }
+
+    @Override
+    public void resourceChanged(IResourceChangeEvent event) {
+        synchronized (needHotCodeReplace) {
+            needHotCodeReplace.set(true);
+        }
+
+        if (event.getType() == IResourceChangeEvent.POST_BUILD) {
+            ChangedClassFilesVisitor visitor = getChangedClassFiles(event);
+            if (visitor != null) {
+                List<IResource> resources = visitor.getChangedClassFiles();
+                List<String> fullyQualifiedName = visitor.getQualifiedNamesList();
+                if (resources != null && !resources.isEmpty()) {
+                    doHotCodeReplace(resources, fullyQualifiedName);
+                } else {
+                    synchronized (needHotCodeReplace) {
+                        needHotCodeReplace.set(false);
+                        needHotCodeReplace.notifyAll();
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public CompletableFuture<List<String>> completed() {
+        synchronized (needHotCodeReplace) {
+            try {
+                needHotCodeReplace.wait(100);
```

Comment:
what's the purpose for wait twice?

Reply:
Long story. And it is the price we have to pay if use in a functional programming.